### PR TITLE
Allow sending RTC with consentState policy adherence

### PR DESCRIPTION
**Pull request by bradfrizzell**
See https://github.com/ampproject/amphtml/issues/15588

### CODE REVIEW

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 1]
```diff
@@ -160,20 +166,94 @@ export class RealTimeConfigManager {
...
     this.handleRtcForVendorUrls(customMacros);
-    return Promise.all(this.promiseArray_);
+    return this.promiseArray_.length ? Promise.all(this.promiseArray_) :
+      null;
+  }
+
+  /**
+   * Returns whether a given callout object is valid to send an RTC request
+   * to, for the given consentState.
+   * @param {Object|string} callout
```
Comment (glevitzky) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: Could you add a return type to the jsdoc here?

File: build-system/amp.extern.js [hunk 2]
```diff
@@ -445,3 +445,10 @@ AMP.require;
  * @typedef {function(number, boolean):?|function(number):?}
  */
 var TransitionDef;
+
+const CONSENT_POLICY_STATE = {
```
Comment (glevitzky) [inline build-system/amp.extern.js]: Please add `/** @enum{number} ... `.

File: extensions/amp-a4a/0.1/real-time-config-manager.js (hunk 1 above)
Comment (bradfrizzell) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: done

File: build-system/amp.extern.js (hunk 2 above)
Comment (bradfrizzell) [reply build-system/amp.extern.js]: done

File: build-system/amp.extern.js [hunk 3]
```diff
@@ -445,3 +445,14 @@ AMP.require;
  * @typedef {function(number, boolean):?|function(number):?}
  */
 var TransitionDef;
+
+/**
+ * See src/consent-state.js for more information.
+ * @enum {number}
+ */
+const CONSENT_POLICY_STATE = {
```
Comment (keithwrightbos) [inline build-system/amp.extern.js]: Given this is now externed can we delete src/consent-state.js (and carry-over documentation)?

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 4]
```diff
@@ -160,20 +166,95 @@ export class RealTimeConfigManager {
...
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: calloutConfig sounds like a better name IMO

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 5]
```diff
@@ -160,20 +166,95 @@ export class RealTimeConfigManager {
...
+   * @return {boolean}
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: @visibleForTesting

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 6]
```diff
@@ -160,20 +166,95 @@ export class RealTimeConfigManager {
...
+   */
+  isValidCalloutForConsentState(callout) {
+    if (!isObject(callout) || !callout['sendRegardlessOfConsentState']) {
+      return false;
+    }
+
+    const sendRegardlessOfConsentState =
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: switch (typeof callout.sendRegardlessOfConsentState) {
   case 'boolean':
      return callout.sendRegardlessOfConsentState;
   case 'Array':
      return sendRegardlessOfConsentState.includes(this.consentState_);
   default:
      user().warn(TAG, 'WHAT?! ...');
      return false;
}

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 7]
```diff
@@ -160,20 +166,95 @@ export class RealTimeConfigManager {
...
+   *     urls: ['https://www.rtc.example/example',
+   *            {url: 'https://www.rtcSite2.example/example',
+   *             sendRegardlessOfConsentState: ['UNKNOWN']}]
+   *    }
+   * and the consentState is CONSENT_POLICY_STATE.UNKNOWN,
+   * then this method call would clear the callouts to vendorB, and to the first
+   * custom URL.
+   */
+  modifyRtcConfigForConsentStateSettings() {
+    if (!(this.consentState_ == CONSENT_POLICY_STATE.INSUFFICIENT ||
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: Duplicate code?  Can you just call isValidCalloutForConsentState with the global value?  Also what if global sendRegardlessOfConsentState is false but individual callout is true (or array with value)?

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 8]
```diff
@@ -226,12 +308,21 @@ export class RealTimeConfigManager {
             this.buildErrorResponse_(
                 RTC_ERROR_ENUM.UNKNOWN_VENDOR, vendor, errorReportingUrl));
       }
+      // There are two valid configurations of the vendor object.
+      // It can either be an object of macros mapping string to string,
+      // or it can be an object with sub-objects, one of which can be
+      // 'macros'. This is for backwards compatability.
+      const vendorMacros =
+            this.rtcConfig_.vendors[vendor]['macros'] &&
+            isObject(this.rtcConfig_.vendors[vendor]['macros']) ?
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: Do you need the first check here?  I would assume isObject returns false if undefined.

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 9]
```diff
@@ -160,20 +166,88 @@ export class RealTimeConfigManager {
...
       return;
     }
+    this.consentState_ = consentState;
+    this.modifyRtcConfigForConsentStateSettings();
     customMacros = this.assignMacros(customMacros);
     this.rtcStartTime_ = Date.now();
     this.handleRtcForCustomUrls(customMacros);
     this.handleRtcForVendorUrls(customMacros);
-    return Promise.all(this.promiseArray_);
+    return this.promiseArray_.length ? Promise.all(this.promiseArray_) :
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: I would assume that Promise.all empty array is equivalent to Promise.resolve... Why return null?

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 10]
```diff
@@ -160,20 +166,88 @@ export class RealTimeConfigManager {
...
+    if (isArray(sendRegardlessOfConsentState)) {
+      for (let i = 0; i < sendRegardlessOfConsentState.length; i++) {
+        if (this.consentState_ ==
+            CONSENT_POLICY_STATE[sendRegardlessOfConsentState[i]]) {
+          return true;
+        }
+      }
+      return false;
+    }
+    user().warn(TAG, 'Invalid value for sendRegardlessOfConsentState');
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: Include callout info?

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 11]
```diff
@@ -160,20 +166,88 @@ export class RealTimeConfigManager {
...
+    }
+
+    if (typeof sendRegardlessOfConsentState == 'boolean') {
+      return sendRegardlessOfConsentState;
+    }
+
+    if (isArray(sendRegardlessOfConsentState)) {
+      for (let i = 0; i < sendRegardlessOfConsentState.length; i++) {
+        if (this.consentState_ ==
+            CONSENT_POLICY_STATE[sendRegardlessOfConsentState[i]]) {
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: We could warn here when a constent state given does not exist in the enum

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 12]
```diff
@@ -160,20 +166,88 @@ export class RealTimeConfigManager {
...
+      return;
+    }
+
+    const urls = [];
+    for (let i = 0; i < this.rtcConfig_.urls.length; i++) {
+      if (this.isValidCalloutForConsentState(this.rtcConfig_.urls[i])) {
+        urls.push(this.rtcConfig_.urls[i]);
+      }
+    }
+    this.rtcConfig_.urls = urls;
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: this.rtcConfig_.urls = this.rtcConfig_.urls.filter(url => this.isValidCalloutForConsentState(url));

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 13]
```diff
@@ -160,20 +166,88 @@ export class RealTimeConfigManager {
...
+   * then this method call would clear the callouts to vendorB, and to the first
+   * custom URL.
+   */
+  modifyRtcConfigForConsentStateSettings() {
+    if (!(this.consentState_ == CONSENT_POLICY_STATE.INSUFFICIENT ||
+          this.consentState_ == CONSENT_POLICY_STATE.UNKNOWN)) {
+      return;
+    }
+
+    if (this.isValidCalloutForConsentState(this.rtcConfig_)) {
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: So global overrides the value on a per callout/vendor level?  Image pub has:

{vendors: {foo: {sendRegardlessOfConsentState: ['unknown']}}, sendRegardlessOfConsentState: true}

If consent state is insufficient should we call out to vendor foo?  I would argue no.

File: extensions/amp-a4a/0.1/real-time-config-manager.js (hunk 12 above)
Comment (bradfrizzell) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: done

File: extensions/amp-a4a/0.1/real-time-config-manager.js (hunk 10 above)
Comment (bradfrizzell) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: done

File: extensions/amp-a4a/0.1/real-time-config-manager.js (hunk 9 above)
Comment (bradfrizzell) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: I thought the same thing but I was having issues in testing where that was not the case

File: extensions/amp-a4a/0.1/real-time-config-manager.js (hunk 11 above)
Comment (bradfrizzell) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: done

Comment (keithwrightbos) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: Could do a switch here:

switch (sendRegardlessOfConsentState[i]) {
    case this.consentState_:
        return true;
    case undefined:
        Log... break;
}

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 14]
```diff
@@ -160,20 +166,85 @@ export class RealTimeConfigManager {
...
-    return Promise.all(this.promiseArray_);
+    return this.promiseArray_.length ? Promise.all(this.promiseArray_) :
+      null;
+  }
+
+  /**
+   * Returns whether a given callout object is valid to send an RTC request
+   * to, for the given consentState.
+   * @param {Object|string} calloutConfig
+   * @param {boolean=} optIsGloballyValid
```
Comment (keithwrightbos) [inline extensions/amp-a4a/0.1/real-time-config-manager.js]: You could make this not optional and just pass false when calling on the global object

File: extensions/amp-a4a/0.1/real-time-config-manager.js (hunk 9 above)
Comment (keithwrightbos) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: This still surprises me... r u sure?

**Pull request merged**